core/src/main/scala/kafka/network/RequestChannel.scala

```diff
@@ -37,6 +37,7 @@ import org.apache.log4j.Logger
 import scala.reflect.ClassTag
 
 object RequestChannel extends Logging {
+  val PrioritizedApiKeys = Set(ApiKeys.LEADER_AND_ISR.id, ApiKeys.UPDATE_METADATA_KEY.id, ApiKeys.STOP_REPLICA.id)
```

Comment:
I remember we discussed this before, but can you remind me why ControlledShutdownRequest is not prioritized?

Reply:
Yeah you commented on this in an above review a long time ago. Joel's reason is why:
https://github.com/apache/kafka/pull/2181#discussion_r90331060

The PR tries to keep request prioritization simple and only prioritize controller-to-broker requests.

Reply:
Hmm, is prioritizing just the controller requests enough? From a client's perspective, it wants not only the metadata to be up to date, but also other things like all live replicas to be in sync. Otherwise, some produce requests with acks=all and min.isr could fail even with up-to-date metadata. So, shouldn't we be prioritizing replica fetch requests as well?

Reply:
@junrao The original idea was to separate the control plain and the data plain so the state transition in the cluster can be propagated faster. Treating the replication traffic differently from the user traffic could be useful. But I feel it is more of a quota problem than a priority issue. If we prioritize the replica fetch request, it might be possible that no client requests can be served because there are always replication fetch in the queue.

Reply:
Hmm, if you can use quota to solve the replication issue, can't you use it to solve the controller request issue? Basically, broker internal communication won't be throttled and client traffic will be throttled.

Also, if the issue is too many requests from the controller, Onur's patch in KAFKA-5642 will reduce the controller request rate during a controlled shutdown significantly.

Reply:
It seems the use case for Quota is mostly for throughput control. And what we are trying to achieve is to improve state change latency. 

The issue we are trying to solve is that when a broker has many requests queued up and moving slowly, the controller requests handling are delayed by long. This may cause a long period of inconsistent state within the cluster because some of the brokers may see the state change much later than the other brokers. It is not only for controlled shutdown but would help with other admin operations in a busy cluster as well. In this case, we actually want to improve the state propagation latency, quota may indirectly help but will not work as well as the prioritized requests.